packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data/generate-property-rows-from-entity/generate-property-row-recursively.ts

```diff
@@ -0,0 +1,103 @@
+import { isPlainObject } from "lodash";
+import {
+  getPropertyTypesByBaseUri,
+  RootEntityAndSubgraph,
+} from "../../../../../../../../../lib/subgraph";
+import { PropertyRow } from "../../types";
+import { getDataTypesOfPropertyType } from "./get-data-types-of-property-type";
+
+/**
+ * This function generates property row data,
+ * and calls itself again for each nested property. Then puts results of these recursive calls into `children` array
+ *
+ * @param properties
+ * Properties object for current depth. On entity level, it starts from `entity.properties`
+ *
+ * @param propertyTypeBaseUri
+ * Name of the specific property inside `properties` object
+ *
+ * @param propertyKeyChain
+ * An array of `propertyTypeBaseUri`'s to store the path to each nested property.
+ * ```js
+ * // Example: `propertyKeyChain` of `c`
+ * properties = { a: { b: { c: "John" } } };
+ * propertyKeyChain = ["a", "b", "c"]
+ * ```
+ *
+ * @param rootEntityAndSubgraph
+ * An object storing root entity & subgraph of that entity
+ *
+ * @param requiredPropertyTypes
+ * An array of `propertyTypeBaseUri`'s.
+ * This is used to check if a property should be shown as `required` or not
+ *
+ * @param depth
+ * Depth of the property. For properties at root, depth starts from `0`
+ *
+ * @returns property row (and nested rows as `children` if it's a nested property)
+ */
+export const generatePropertyRowRecursively = (
+  properties: any,
+  propertyTypeBaseUri: string,
+  propertyKeyChain: string[],
```

Comment:
I don't think it can be just `parentPath`, and it's not only necessary for the recursive implementation. It's also used at `onCellEdited`, and we need to store each key with the correct order to be able to update the nested properties later on.

Reply:
So instead of storing an array like `["company", "ceo", "name"]`, we could store a string like`"company-ceo-name"` (probably gonna need a seperator string different than '-', because the string should never be allowed to be used in the keys, otherwise we cannot use string.split() to seperate them by the seperator). But I don't see a benefit of transforming this to a string, and splitting/merging it in multiple places.

Reply:
> we need to store each key with the correct order to be able to update the nested properties later on.

lodash has a deep set function that uses `dot` notion

Reply:
There's a diff here that explains what I'm arguing for here

https://github.com/hashintel/hash/pull/1266#discussion_r1012104066

Reply:
but do we have a guarantee about `propertyTypeBaseUri`'s never including a `dot`?

Reply:
I think that's a fair point and a good argument for keeping the path as an array – I think you could still just call it `path` though